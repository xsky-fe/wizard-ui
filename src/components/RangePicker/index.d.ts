import { RangePickerProps } from '../../interface';

export default function RangePicker(props: RangePickerProps): JSX.Element;