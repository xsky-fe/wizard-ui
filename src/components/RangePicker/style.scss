@import '../../style/variables.scss';

.RangePicker {
  display: inline-block;
  min-width: 366px;
  width: auto;
  & > input {
    text-align: center;
    border-style: none;
    &:focus {
      outline: 0;
    }
  }
  &__duration {
    border-bottom: 1px solid #e9e9e9;
    overflow: hidden;
    margin-bottom: 10px;
    button {
      float: left;
    }
  }
  &:hover span.icon-close {
    visibility: visible;
  }
  span.glyphicon-calendar{
    line-height: unset;
    color: #ccc;
  }
  span.icon-close {
    visibility: hidden;
    font-size: 12px;
    cursor: pointer;
    vertical-align: inherit;
  }
}

.rc-calendar-year-panel-selected-cell .rc-calendar-year-panel-year,
.rc-calendar-decade-panel-selected-cell .rc-calendar-decade-panel-decade,
.rc-calendar-month-panel-selected-cell .rc-calendar-month-panel-month {
  background: $purple-normal;
}
.rc-calendar-prev-month-btn:hover,
.rc-calendar-next-month-btn:hover,
.rc-calendar-prev-year-btn:hover,
.rc-calendar-next-year-btn:hover,
.rc-calendar-year-select:hover,
.rc-calendar-month-select:hover,
.rc-calendar-day-select:hover,
.rc-calendar-decade-panel-header > a:hover,
.rc-calendar-month-panel-header > a:hover,
.rc-calendar-year-panel-header > a:hover,
li.rc-time-picker-panel-select-option-selected {
  color: $purple-normal;
}
.rc-calendar-month-panel-cell .rc-calendar-month-panel-month:hover,
.rc-calendar-year-panel-year:hover,
.rc-calendar-decade-panel-decade:hover,
.rc-time-picker-panel-select li:hover,
.rc-calendar-date:hover {
  background: $purple-lighter-2 !important;
}
.rc-calendar-range .rc-calendar-in-range-cell,
li.rc-time-picker-panel-select-option-selected {
  background: $purple-lighter-3;
}

.rc-calendar-date-panel {
  border-top: 3px solid $purple-normal;
}
.rc-calendar-today .rc-calendar-date {
  border: 1px solid $purple-normal;
}
.rc-calendar-selected-date .rc-calendar-date {
    background: transparent;
    color: $gray-medium-13;
    &:hover {
      background: $purple-lighter-2;
    }
}
.rc-calendar-ok-btn,
.rc-calendar-selected-day .rc-calendar-date {
  background: $purple-normal;
  color: $white;
}
.rc-calendar-time-picker-btn {
  color: $purple-normal;
  &:hover {
    color: $purple-normal;
  }
  &.rc-calendar-time-picker-btn-disabled {
    color: #bbb;
  }
}
.rc-calendar-ok-btn {
  padding: 5px 15px;
  border-radius: 3px;
  margin-right: 12px;
  &:hover {
    color: $white;
  }
  &.rc-calendar-ok-btn-disabled {
    cursor: not-allowed;
    background: $purple-darker-4
  }
}
.rc-time-picker-panel-select ul {
  max-height: 216px;
}
.rc-calendar-time-picker {
  border-bottom: 1px solid #e9e9e9;
}
.rc-calendar-range .rc-calendar-footer-btn {
  padding: 10px 0;
}
