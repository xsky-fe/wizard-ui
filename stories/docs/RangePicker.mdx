import { Canvas, Meta, ArgTypes } from '@storybook/blocks';

import * as RangePickerStories from '../RangePicker.stories.tsx';

<Meta of={RangePickerStories} />

# RangePicker
<em><strong>(wizard-ui自定义组件)</strong></em>
<p>2023-6-19 [peishan](https://github.xsky.com/peishan)</p>

选定开始时间和结束时间输出时间区间。基于 [rc-calendar](https://github.com/react-component/calendar) 和[rc-time-picker](https://github.com/react-component/time-picker)开发。

## 基本用法
1. 开始时间小于结束时间。
2. 支持选择日期同时支持选择时间，精确到秒级别。
3. 支持中文、英文。默认可以根据浏览器的本地存储变量LOCALE决定切换。
4. 支持区间（3 小时内、1 天内、1 周内、1 月内、3 月内）选择，选择区间会立即关闭选择面板。

## 代码演示
### 默认展示
<Canvas of={RangePickerStories.Basic}/>
### 设置默认时间区间
<Canvas of={RangePickerStories.DefaultTimes}/>
### 禁用当天后的日期
<Canvas of={RangePickerStories.disabledAfterToday}/>
### 禁用当天后的日期及时间
<Canvas of={RangePickerStories.disabledAfterTodayTime}/>
### 设置 `showDuration` 为 `false` 隐藏直接选择区间
<Canvas of={RangePickerStories.HideDuration}/>

## API
<ArgTypes />