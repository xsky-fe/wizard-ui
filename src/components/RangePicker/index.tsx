import * as React from 'react';
import { Button } from 'react-bootstrap';
import * as PropTypes from 'prop-types';
import Picker from 'rc-calendar/lib/Picker';
import RangeCalendar from 'rc-calendar/lib/RangeCalendar';
import TimePickerPanel from 'rc-time-picker/lib/Panel';
import 'rc-calendar/assets/index.css';
import 'rc-time-picker/assets/index.css';
import CALENDAR_ZH from 'rc-calendar/lib/locale/zh_CN';
import CALENDAR_EN from 'rc-calendar/lib/locale/en_US';
import moment, { Moment, DurationInputArg1, DurationInputArg2 } from 'moment';
import { RangePickerProps, RangePickerState } from '../../interface';
import isEmpty from 'lodash/isEmpty';
import Icon from '../Icon';
import './style.scss';
const ranges = [
  {
    title: '3 小时内',
    value: '3 hours',
  },
  {
    title: '1 天内',
    value: '1 day',
  },
  {
    title: '1 周内',
    value: '1 week',
  },
  {
    title: '1 月内',
    value: '1 month',
  },
  {
    title: '3 月内',
    value: '3 months',
  },
];

function disabledDate(current: Moment) {
  if (!current) return false;
  const date = moment();
  return current.isAfter(date); // can not select days after today
}

export default class RangePicker extends React.PureComponent<RangePickerProps, RangePickerState> {
  static defaultProps = {
    disabled: false,
    showDuration: true,
    format: 'YYYY-MM-DD HH:mm:ss',
    defaultValue: [],
  };
  static propTypes = {
    /**
     * 禁用
     */
    disabled: PropTypes.bool,
    /**
     * 改变时间区间会触发 onChange 执行
     */
    onChange: PropTypes.func,
    /**
     * 点击确定会触发
     */
    onOk: PropTypes.func,
    /**
     * moment 格式化
     */
    format: PropTypes.string,
    /**
     * 是否展示直接选择区间
     */
    showDuration: PropTypes.bool,
    defaultValue: PropTypes.array,
  };
  constructor(props: RangePickerProps) {
    super(props);
    const value = props.value || props.defaultValue || [];
    this.state = {
      value,
      open: false,
    };
  }
  get lang() {
    const lang = window.localStorage.getItem('LOCALE')
    return lang && /en/.test(lang) ? 'en' : 'zh';
  }
  get locale() {
    return this.lang === 'en' ? CALENDAR_EN : CALENDAR_ZH;
  }
  get startPlaceholder() {
    /* eslint-disable-next-line */
    return this.lang === 'en' ? 'Start time' : '开始时间';
  }
  get endPlaceholder() {
    /* eslint-disable-next-line */
    return this.lang === 'en' ? 'End time' : '结束时间';
  }
  get seperator() {
    /* eslint-disable-next-line */
    return this.lang === 'en' ? 'To' : '至';
  }
  formatValue = (value: Moment) => {
    const { format } = this.props;
    return (value && value.format(format)) || '';
  };
  handleOuterChange = (value: Moment[]) => {
    const { onChange } = this.props;
    if (onChange) {
      onChange(value, [this.formatValue(value[0]), this.formatValue(value[1])]);
    }
  };
  handleChange = (value: Moment[]) => {
    this.handleOuterChange(value);
    this.setState({ value });
  };
  handleRangeClick = (value: any) => {
    const { onOk } = this.props;
    const end: Moment = moment(new Date());
    const num: DurationInputArg1 = value.split(' ')[0];
    const unit: DurationInputArg2 = value.split(' ')[1];
    const start = moment(new Date()).subtract(num, unit);
    const duration = [start, end];
    this.handleOuterChange(duration);
    if (onOk) {
      onOk(duration);
    }
    this.setState({
      value: duration,
      open: false,
    });
  };
  handleOpenChange = (open: boolean) => {
    const { onOpenChange } = this.props;
    this.setState({ open });
    if (onOpenChange) {
      onOpenChange(open);
    }
  };
  clearSelection = (e: React.MouseEvent) => {
    e.preventDefault();
    e.stopPropagation();
    const { onOk } = this.props;
    this.setState({ value: [] });
    this.handleChange([]);
    if (onOk) {
      onOk([]);
    }
  };
  renderFooter = () => {
    if (!this.props.showDuration) {
      return;
    }
    return (
      <div className="RangePicker__duration">
        {ranges.map(({ title, value }) => {
          return (
            <Button bsStyle="link" key={value} onClick={() => this.handleRangeClick(value)}>
              {this.lang === 'en' ? value : title}
            </Button>
          );
        })}
      </div>
    );
  };
  renderCalendar = () => {
    const { onOk, format } = this.props;
    return (
      <RangeCalendar
        seperator={this.seperator}
        dateInputPlaceholder={[this.startPlaceholder, this.endPlaceholder]}
        locale={this.locale}
        format={format}
        onOk={onOk}
        disabledDate={disabledDate}
        timePicker={<TimePickerPanel />}
        renderFooter={this.renderFooter}
        showToday={false}
      />
    );
  };
  renderClearIcon = () => {
    if (this.state.value.every(isEmpty)) {
      return;
    }
    return <Icon type="close" onClick={this.clearSelection} />;
  };
  render() {
    const { disabled } = this.props;
    return (
      <Picker
        value={this.state.value}
        disabled={disabled}
        onChange={this.handleChange}
        open={this.state.open}
        onOpenChange={this.handleOpenChange}
        animation="slide-up"
        calendar={this.renderCalendar()}
      >
        {({ value }: { value?: any }) => {
          const [start, end] = value;
          return (
            <span className="RangePicker form-control">
              <span className="glyphicon glyphicon-calendar" />
              <input
                placeholder={this.startPlaceholder}
                readOnly
                value={this.formatValue(start)}
                tabIndex={-1}
              />
              <span>{this.seperator}</span>
              <input
                placeholder={this.endPlaceholder}
                readOnly
                value={this.formatValue(end)}
                tabIndex={-1}
              />
              {this.renderClearIcon()}
            </span>
          );
        }}
      </Picker>
    );
  }
}
