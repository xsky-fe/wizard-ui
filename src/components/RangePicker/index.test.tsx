import React from 'react';
import { mount } from 'enzyme';
import RangePicker from './index';
import renderer from 'react-test-renderer';
import moment, { Moment } from 'moment';

function onChange(date: Moment[], dateString: string[]) {
  console.log('onChange', date, dateString);
}
function onOk(date: Moment[]) {
  console.log('onOk', date);
}
function onOpenChange(open: boolean) {
  console.log('open', open);
}

describe('RangePicker', () => {
  it('render with agreed class', () => {
    const picker = mount(<RangePicker />);
    const node = picker.find('.RangePicker');
    // class RangePicker 的节点有 1个
    expect(node.length).toBe(1);
    node.simulate('click');
    // 点击后立即更新
    picker.update();
    // 展开日期选择框 dom 节点 div.rc-calendar-picker 有一个
    expect(picker.find('div.rc-calendar-picker').length).toBe(1);
  });
  it('render show primary without props inject', () => {
    const picker = renderer.create(<RangePicker />).toJSON();
    expect(picker).toMatchSnapshot();
  });
  it('render show with en lang', () => {
    const picker = renderer.create(<RangePicker lang="en" />).toJSON();
    expect(picker).toMatchSnapshot();
  });
  it('click show time panel', () => {
    const picker = mount(
      <RangePicker defaultValue={[moment('2019-02-16 18:01:57'), moment('2019-02-20 18:01:57')]} />,
    );
    const node = picker.find('.RangePicker');
    node.simulate('click');
    picker
      .find('a.rc-calendar-time-picker-btn')
      .at(0)
      .simulate('click');
    expect(picker.find('.rc-calendar-time-picker').length).toBe(2);
  });
  it('clear choosed duration', () => {
    const picker = mount(
      <RangePicker
        defaultValue={[moment('2019-02-16 18:01:57'), moment('2019-02-20 18:01:57')]}
        onOk={onOk}
      />,
    );
    const node = picker.find('.RangePicker');
    node.simulate('mouseenter');
    picker
      .find('.icon-close-line')
      .at(0)
      .simulate('click');
    const inputs = picker.find('input');
    // 开始时间
    expect(inputs.at(0).props().value).toBe('');
    // 结束时间
    expect(inputs.at(1).props().value).toBe('');
  });
  it('hide duration', () => {
    const picker = mount(<RangePicker showDuration={false} />);
    const node = picker.find('.RangePicker');
    node.simulate('click');
    expect(picker.find('.RangePicker__duration').length).toBe(0);
  });
  it('3 hours duration', () => {
    const picker = mount(
      <RangePicker onChange={onChange} onOk={onOk} onOpenChange={onOpenChange} />,
    );
    const node = picker.find('.RangePicker');
    node.simulate('click');
    picker
      .find('.RangePicker__duration button')
      .at(0)
      .simulate('click');
    const inputs = picker.find('input');
    // 开始时间
    expect(inputs.at(0).props().value).not.toBe('');
    // 结束时间
    expect(inputs.at(1).props().value).not.toBe('');
  });
});
