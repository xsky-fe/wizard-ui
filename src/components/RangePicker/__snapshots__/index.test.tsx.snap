// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangePicker render show primary without props inject 1`] = `
<span
  className="RangePicker form-control"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <span
    className="glyphicon glyphicon-calendar"
  />
  <input
    placeholder="开始时间"
    readOnly={true}
    tabIndex={-1}
    value=""
  />
  <span>
    至
  </span>
  <input
    placeholder="结束时间"
    readOnly={true}
    tabIndex={-1}
    value=""
  />
</span>
`;

exports[`RangePicker render show with en lang 1`] = `
<span
  className="RangePicker form-control"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <span
    className="glyphicon glyphicon-calendar"
  />
  <input
    placeholder="开始时间"
    readOnly={true}
    tabIndex={-1}
    value=""
  />
  <span>
    至
  </span>
  <input
    placeholder="结束时间"
    readOnly={true}
    tabIndex={-1}
    value=""
  />
</span>
`;
